client/my-sites/plans-grid/components/plan-type-selector/hooks/use-term-vise-max-dicounts.ts

```diff
@@ -0,0 +1,97 @@
+import {
+	PlanSlug,
+	URL_FRIENDLY_TERMS_MAPPING,
+	UrlFriendlyTermType,
+	getPlanMultipleTermsVariantSlugs,
+	getPlanSlugForTermVariant,
+	isMonthly,
+	isWpComPlan,
+	isWpcomEnterpriseGridPlan,
+} from '@automattic/calypso-products';
+import { UsePricingMetaForGridPlans } from 'calypso/my-sites/plans-grid/hooks/npm-ready/data-store/use-grid-plans';
+
+export default function useTermViseMaxDiscounts(
+	plans: PlanSlug[],
+	urlFriendlyTerms: UrlFriendlyTermType[] = [],
+	usePricingMetaForGridPlans: UsePricingMetaForGridPlans
+): Record< UrlFriendlyTermType, number > {
+	const termDefinitionsMapping = urlFriendlyTerms.map( ( urlFriendlyTerm ) => ( {
+		urlFriendlyTerm,
+		term: URL_FRIENDLY_TERMS_MAPPING[ urlFriendlyTerm ],
+	} ) );
+	const terms = termDefinitionsMapping.map( ( { term } ) => term );
+	if ( terms.length !== urlFriendlyTerms.length ) {
+		throw new Error( 'Invalid term passed in:' + JSON.stringify( urlFriendlyTerms ) );
+	}
+
+	const allRelatedPlanSlugs = plans
+		.filter( ( planSlug ) => ! isWpcomEnterpriseGridPlan( planSlug ) )
+		.map( ( plan ) => getPlanMultipleTermsVariantSlugs( plan, terms ) )
+		.flat()
+		.filter( Boolean )
+		.filter( isWpComPlan ) as PlanSlug[];
+
+	const wpcomMonthlyPlanSlugs = allRelatedPlanSlugs.filter( isMonthly );
+	const wpcomNonMonthlyPlans = allRelatedPlanSlugs.filter(
+		( planSlug ) => ! isMonthly( planSlug )
+	);
+
+	const monthlyPlansPricing = usePricingMetaForGridPlans( {
+		planSlugs: wpcomMonthlyPlanSlugs,
+		withoutProRatedCredits: true,
+		storageAddOns: null,
+	} );
+	const nonMonthlyPlansPricing = usePricingMetaForGridPlans( {
+		planSlugs: wpcomNonMonthlyPlans,
+		withoutProRatedCredits: true,
+		storageAddOns: null,
+	} );
+
+	const getTermInMonths = ( term: UrlFriendlyTermType ): number => {
+		switch ( term ) {
+			case '3yearly':
+				return 36;
+			case '2yearly':
+				return 24;
+			case 'yearly':
+				return 12;
+			case 'monthly':
+			default:
+				return 1;
+		}
+	};
+
+	const termViseMaxDiscount: Record< UrlFriendlyTermType, number > = {} as Record<
+		UrlFriendlyTermType,
+		number
+	>;
+	termDefinitionsMapping.forEach( ( termMapping ) => {
```

Comment:
`NB:`

We agree that the density of transformations / operations is surprising, but I suppose when we give it more thought, calculating all possible discounts for all variations of each plan ( personal, premium, business, etc. ) from scratch could be a bit complex.

@jdc91 brought up a good point -- perhaps we could create an abstraction for the discount calculation for a single term. It's not something I'm convinced that we should immediately address since this is an experiment. I _do_ believe that we also agree that if / when we do want to release this to a general audience, refactoring this block of code becomes a blocker.